### ISSUE TITLE
Regarding the issue of reproducing the results of llava-1.5

### ISSUE THREAD

**Issue by glory6-CYL** [zjunlp/Deco#6]
Thank you very much for opening up such excellent work, but I encounter some problems as follow:
When I use the ./opera_log/llava-1.5/ours.jsonl to run the chair.py,the result is as following picture which is different from your paper.Could you please  teach me how to get the correct result?

![Image](https://github.com/user-attachments/assets/92de184d-6c83-4c98-ba45-9f6153144d4c)

And when I use the llava-1.5 which is called models--liuhaotian--llava-v1.5-7b, the result is:

![Image](https://github.com/user-attachments/assets/3f5b8192-7267-4fa5-a29f-912ad3ed2911)
Is it problem because of the model?Could you please tell me which llava-v1.5-7b did you use?
Thank you very much!

**Comment by sunnychenxiwang** [comment]
For the first question, we refer to the results from the [OPERA](https://arxiv.org/abs/2311.17911), which reports CHAIRs = 44.6 and CHAIRi = 12.8 on LLava-1.5. These results are not significantly different from actual experimental outcomes.

For the second question, the model we used is indeed models--liuhaotian--llava-v1.5-7b. The discrepancies may be due to differences in prompts, image data, or decoding strategies. If possible, could you provide these details?

**Comment by glory6-CYL** [comment]
About running the chair_llava.py, I just change the eval_model() and the parser as follow(I have changed the following code to run on a single GPU and bolded the part where I changed the path):
def eval_model(args):
    # 设置单 GPU 设备
    device = "cuda:0"

    # 设置实验文件夹和日志
    os.makedirs(args.log_path, exist_ok=True)
    model_string_name = args.model_path.split("/")[-1]
    experiment_dir = args.log_path
    os.makedirs(experiment_dir, exist_ok=True)
    logger = setup_logger(experiment_dir)  # 使用内置日志替代
    logger.info(f"Experiment directory created at {experiment_dir}")

    # 加载模型
    disable_torch_init()
    model_path = os.path.expanduser(args.model_path)
    model_name = get_model_name_from_path(model_path)
    tokenizer, model, image_processor, context_len = load_pretrained_model(
        model_path, args.model_base, model_name, device=device
    )
    model.to(device)

    answers_file = os.path.expanduser(args.answers_file)
    os.makedirs(os.path.dirname(answers_file), exist_ok=True)

    with open("./opera_log/llava-1.5/ours.jsonl", "r", encoding="utf-8") as f:
        for _, data_line in tqdm(enumerate(f.readlines()), total=500):
            line = json.loads(data_line)
            idx = line["image_id"]
            image_file = "./val2014/val2014/COCO_val2014_" + str(idx).zfill(12) + ".jpg"
            qs = "Please describe this image in detail."
            cur_prompt = qs

            if model.config.mm_use_im_start_end:
                qs = DEFAULT_IM_START_TOKEN + DEFAULT_IMAGE_TOKEN + DEFAULT_IM_END_TOKEN + '\n' + qs
            else:
                qs = DEFAULT_IMAGE_TOKEN + '\n' + qs

            conv = conv_templates[args.conv_mode].copy()
            conv.append_message(conv.roles[0], qs)
            conv.append_message(conv.roles[1], None)
            prompt = conv.get_prompt()

            input_ids = tokenizer_image_token(prompt, tokenizer, IMAGE_TOKEN_INDEX, return_tensors='pt').unsqueeze(0).to(device)
            image = Image.open(image_file)
            image_tensor = image_processor.preprocess(image, return_tensors='pt')['pixel_values'][0].to(device)

            stop_str = conv.sep if conv.sep_style != SeparatorStyle.TWO else conv.sep2
            keywords = [stop_str]
            stopping_criteria = KeywordsStoppingCriteria(keywords, tokenizer, input_ids)

            with torch.inference_mode():
                with torch.no_grad():
                    output_dict = model.generate(
                        input_ids,
                        images=image_tensor.unsqueeze(0).half().to(device),
                        do_sample=True if args.temperature > 0 else False,
                        temperature=args.temperature,
                        top_p=args.top_p,
                        num_beams=args.num_beams,
                        max_new_tokens=args.max_new_tokens,
                        return_dict_in_generate=True,
                        output_hidden_states=True,
                        stopping_criteria=[stopping_criteria],
                        use_deco=True,
                        alpha=args.alpha,
                        threshold_top_p=args.threshold_top_p,
                        threshold_top_k=args.threshold_top_k,
                        early_exit_layers=[i for i in range(args.start_layer, args.end_layer)],
                        return_dict=True
                    )

            output_ids = output_dict.sequences
            input_token_len = input_ids.shape[1]
            outputs = tokenizer.batch_decode(
                output_ids[:, input_token_len:], skip_special_tokens=True
            )[0]
            outputs = outputs.strip()

            logger.info(f"[{image_file}]")
            logger.info(f"prompt: {cur_prompt}")
            logger.info(f"text: {outputs}")
            res_dict = {"id": idx, "response": outputs}
            #ans_file.write(json.dumps(res_dict, ensure_ascii=False) + "\n")

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    **parser.add_argument("--model-path", type=str, default=r"C:\Users\PC\.cache\huggingface\hub\models--liuhaotian--llava-v1.5-7b\snapshots\4481d270cc22fd5c4d1bb5df129622006ccd9234")**
    parser.add_argument("--model-base", type=str, default=None)
    **parser.add_argument("--answers-file", type=str, default="C:/Users/PC/Deco/answers/output.jsonl")**
    parser.add_argument("--conv-mode", type=str, default="llava_v1")
    parser.add_argument("--num-chunks", type=int, default=1)
    parser.add_argument("--chunk-idx", type=int, default=0)
    parser.add_argument("--temperature", type=float, default=-1)
    parser.add_argument("--top_p", type=float, default=None)
    parser.add_argument("--top_k", type=int, default=None)
    parser.add_argument("--num_beams", type=int, default=1)
    parser.add_argument("--max_new_tokens", type=int, default=512)
    **parser.add_argument("--log_path", type=str, default="C:/Users/PC/Deco/logs")**
    parser.add_argument("--batch_size", type=int, default=1)
    parser.add_argument("--alpha", type=float, default=0.6)
    parser.add_argument("--threshold_top_p", type=float, default=0.9)
    parser.add_argument("--threshold_top_k", type=int, default=20)
    parser.add_argument("--start_layer", type=int, default=20)
    parser.add_argument("--end_layer", type=int, default=29)
    parser.add_argument("--seed", type=int, default=42)

    args = parser.parse_args()
    set_seed(args.seed)
    eval_model(args)

After that, I processed the log file:

![Image](https://github.com/user-attachments/assets/f67a9f1f-03e6-485e-a2a8-0c3ff5b038db)
into:

![Image](https://github.com/user-attachments/assets/3335c572-2a8b-4415-a65b-594f3dd2419c)
After that, I use this jsonl file to run the chair.py, the results are as follow:
{
      "image_id": 226988,
      "caption": "The image depicts a woman sitting at a table, taking a picture of two delicious looking pizzas placed on a plate. She is using a camera to capture the moment, capturing both pizzas in her shot. The table occupies most of the scene, with the woman occupying the left side of the table.",
      "mscoco_hallucinated_words": [],
      "mscoco_gt_words": [
        "dining table",
        "pizza",
        "person"
      ],
      "mscoco_generated_words": [
        "person",
        "dining table",
        "pizza",
        "pizza",
        "dining table",
        "person",
        "dining table"
      ],
      "hallucination_idxs": [],
      "words": [
        "the",
        "image",
        "depict",
        "a",
        "woman",
        "sit",
        "at",
        "a",
        "table",
        ",",
        "take",
        "a",
        "picture",
        "of",
        "two",
        "delicious",
        "look",
        "pizza",
        "place",
        "on",
        "a",
        "plate",
        ".",
        "she",
        "be",
        "use",
        "a",
        "camera",
        "to",
        "capture",
        "the",
        "moment",
        ",",
        "capture",
        "both",
        "pizza",
        "in",
        "her",
        "shot",
        ".",
        "the",
        "table",
        "occupies",
        "most",
        "of",
        "the",
        "scene",
        ",",
        "with",
        "the",
        "woman",
        "occupy",
        "the",
        "left",
        "side",
        "of",
        "the",
        "table",
        "."
      ],
      "metrics": {
        "CHAIRs": 0,
        "CHAIRi": 0.0,
        "Recall": 1.0,
        "Len": 0.0
      }
    },(one of the results)
It's strange that the generated sentence seems to be very similar to the original caption, so I initially doubted if the model training I used already included the Coco dataset. But if you say you are using the same model, I don't know why this situation occurs. If you need it, I can send you this JSON file again.

**Comment by glory6-CYL** [comment]
这是我调用chair_llava.py跑出来的log文件处理得到的结果




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zjunlp/Deco"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年3月5日(星期三) 晚上8:04
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [zjunlp/Deco] Regarding the issue of reproducing the results of llava-1.5 (Issue #6)



   
For the first question, we refer to the results from the OPERA, which reports CHAIRs = 44.6 and CHAIRi = 12.8 on LLava-1.5. These results are not significantly different from actual experimental outcomes.
 
For the second question, the model we used is indeed models--liuhaotian--llava-v1.5-7b. The discrepancies may be due to differences in prompts, image data, or decoding strategies. If possible, could you provide these details?

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***&gt;
  sunnychenxiwang left a comment (zjunlp/Deco#6)
 
For the first question, we refer to the results from the OPERA, which reports CHAIRs = 44.6 and CHAIRi = 12.8 on LLava-1.5. These results are not significantly different from actual experimental outcomes.
 
For the second question, the model we used is indeed models--liuhaotian--llava-v1.5-7b. The discrepancies may be due to differences in prompts, image data, or decoding strategies. If possible, could you provide these details?

**Comment by sunnychenxiwang** [comment]
Could you send the jsonl file to my mail sunnywcx@zju.edu.cn

**Comment by sunnychenxiwang** [comment]
![Image](https://github.com/user-attachments/assets/6e65ae41-dca3-47b5-95fe-346f86555547)
From your results, the generated sentences tend to be relatively short and exhibit some semantic inconsistencies, such as repetition. In our experiments, we considered both the quality of the generated sentences, including semantic coherence and richness, and did not solely pursue the lowest CHAIR metric. You may consider adjusting alpha and selecting different layers to control this.

**Comment by glory6-CYL** [comment]
Thank you for you early reply,I will try it latter!
